R/verb-pivot-longer.R

```diff
@@ -0,0 +1,302 @@
+                                     names_to = "name",
+                                     values_to = "value",
+                                     names_prefix = NULL,
+                                     names_sep = NULL,
+                                     names_pattern = NULL,
+                                     names_ptypes = NULL,
+                                     names_transform = NULL) {
+  cn_data <- colnames(data)
+  data_tmp <- set_names(character(length(cn_data)), cn_data)
+  cols <- tidyselect::eval_select(enquo(cols), data_tmp)
+
+  if (length(cols) == 0) {
+    abort(glue::glue("`cols` must select at least one column."))
+  }
+
+  if (is.null(names_prefix)) {
+    names <- names(cols)
+  } else {
+    names <- gsub(paste0("^", names_prefix), "", names(cols))
+  }
+
+  if (length(names_to) > 1) {
+    if (!xor(is.null(names_sep), is.null(names_pattern))) {
+      abort(glue::glue(
+        "If you supply multiple names in `names_to` you must also supply one",
+        " of `names_sep` or `names_pattern`."
+      ))
+    }
+
+    if (!is.null(names_sep)) {
+      names <- str_separate(names, names_to, sep = names_sep)
+    } else {
+      names <- str_extract(names, names_to, regex = names_pattern)
+    }
+  } else if (length(names_to) == 0) {
+    names <- tibble::new_tibble(x = list(), nrow = length(names))
+  } else {
+    if (!is.null(names_sep)) {
+      abort("`names_sep` can not be used with length-1 `names_to`")
+    }
+    if (!is.null(names_pattern)) {
+      names <- str_extract(names, names_to, regex = names_pattern)[[1]]
+    }
+
+    names <- tibble(!!names_to := names)
+  }
+
+  if (".value" %in% names_to) {
+    values_to <- NULL
+  } else {
+    vctrs::vec_assert(values_to, ptype = character(), size = 1)
+  }
+
+  # optionally, cast variables generated from columns
+  cast_cols <- intersect(names(names), names(names_ptypes))
+  for (col in cast_cols) {
+    names[[col]] <- vctrs::vec_cast(names[[col]], names_ptypes[[col]])
+  }
+
+  # transform cols
+  coerce_cols <- intersect(names(names), names(names_transform))
+  for (col in coerce_cols) {
+    f <- as_function(names_transform[[col]])
+    names[[col]] <- f(names[[col]])
+  }
+
+  out <- tibble(.name = names(cols))
+  out[[".value"]] <- values_to
+  out <- vctrs::vec_cbind(out, names)
+  out
+}
+
+dbplyr_pivot_longer_spec <- function(data,
+                                     spec,
+                                     names_repair = "check_unique",
+                                     values_drop_na = FALSE,
+                                     values_transform = list()) {
+  spec <- check_spec(spec)
+  # .seq col needed if different input columns are mapped to the same output
+  # column
+  spec <- deduplicate_spec(spec, data)
+
+  id_cols <- syms(setdiff(colnames(data), spec$.name))
+
+  spec_split <- vctrs::vec_split(spec, spec[, -(1:2)])
+  data_long_list <- purrr::map(
+    vctrs::vec_seq_along(spec_split),
+    function(idx) {
```

Comment:
I think this is complicated enough that I'd want to pull it out into a top-level function

Reply:
I instead extracted parts of the function into a top level function so there is no map() called directly. I think it looks quite simple now.